various cross and gobject-introspection fixes

Artturin opened the pull request:
remove unnecessary preInstall

###### Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

Artturin commented on pkgs/development/libraries/gdk-pixbuf/default.nix:
```diff
@@ -108,9 +107,9 @@ stdenv.mkDerivation rec {
+    '' + ''
       # We need to install 'loaders.cache' in lib/gdk-pixbuf-2.0/2.10.0/
-      $dev/bin/gdk-pixbuf-query-loaders --update-cache
+      ${stdenv.hostPlatform.emulator buildPackages} $dev/bin/gdk-pixbuf-query-loaders --update-cache
```
this will be bash on build == host and a emulator otherwise

gobject-introspection already brings in the emulator to the closure

Mindavi commented:
Can we now start setting strictDeps on these too or do you want to defer that?

Mindavi reviewed (approved):
Diff looks good to me. We should take care to not remove the flags in packages where it's not enabled by default, if there are any that don't enable it by default.

Artturin commented:
> Can we now start setting strictDeps on these too or do you want to defer that?

lets do it in a separate pr

i use a command like this to generate the list of attrs and then i input that in to diffing
`git diff --name-only HEAD HEAD~6 | rev | cut -d '/' -f 2 | rev > attrlist`

jtojnar commented on pkgs/development/libraries/json-glib/default.nix:
```diff
@@ -37,22 +36,26 @@ stdenv.mkDerivation rec {
-    "-Dgtk_doc=disabled"
-  ];
+  postPatch = ''
+    substituteInPlace doc/meson.build \
```
Add TODO comment here as well

jtojnar commented on pkgs/development/libraries/atk/default.nix:
```diff
@@ -39,15 +39,13 @@ stdenv.mkDerivation rec {
     fixDarwinDylibNames
   ];
 
+  buildInputs = [ gobject-introspection ];
```
Why add this to `buildInputs`? It should not be a runtime dependency.

Mindavi replied:
libgirepository can be used at runtime, right?

jtojnar replied:
Yes, but ATK should not depend on `libgirepository`.

Mindavi replied:
Then it shouldn't be in buildInputs :)

Mindavi replied:
For some reason it then does require it at runtime, or the meson file is wrong.

jtojnar replied:
Perhaps it needs `native=True` here: https://github.com/mesonbuild/meson/blob/16ed82a62f535475b0eff6acc2d7c34c5265c16f/mesonbuild/modules/gnome.py#L772

Artturin replied:
i applied

```diff
diff --git a/pkgs/development/tools/build-managers/meson/default.nix b/pkgs/development/tools/build-managers/meson/default.nix
index f62bb22314d..b46f918b9c3 100644
--- a/pkgs/development/tools/build-managers/meson/default.nix
+++ b/pkgs/development/tools/build-managers/meson/default.nix
@@ -71,6 +71,11 @@ python3.pkgs.buildPythonApplication rec {
     ./fix-gtkdoc-when-using-multiple-apple-frameworks.patch
   ];
 
+  postPatch = ''
+    substituteInPlace mesonbuild/modules/gnome.py \
+      --replace "(state, 'gobject-introspection-1.0')" "(state, 'gobject-introspection-1.0', native = True)"
+  '';
+
   setupHook = ./setup-hook.sh;
 
   # Meson included tests since 0.45, however they fail in Nixpkgs because they
```

to meson because of some projects using the wrong gi when build == host != target and gobject-introspection was in buildInputs

that fixed them

gobject-introspection still has to be in buildInputs

otherwise

```
json-glib-aarch64-unknown-linux-gnu> /nix/store/s9m5ww62bsyx0jzylndxicvkmdn7x5bj-aarch64-unknown-linux-gnu-binutils-2.38/bin/aarch64-unknown-linux-gnu-ld: skipping incompatible /nix/store/c79djpxhga419drlbf6v850pvpjfqsgl-gobject-introspection-wrapped-1.72.0/lib/libgirepository-1.0.so when searching for -lgirepository-1.0
json-glib-aarch64-unknown-linux-gnu> /nix/store/s9m5ww62bsyx0jzylndxicvkmdn7x5bj-aarch64-unknown-linux-gnu-binutils-2.38/bin/aarch64-unknown-linux-gnu-ld: cannot find -lgirepository-1.0: No such file or directory
json-glib-aarch64-unknown-linux-gnu> /nix/store/s9m5ww62bsyx0jzylndxicvkmdn7x5bj-aarch64-unknown-linux-gnu-binutils-2.38/bin/aarch64-unknown-linux-gnu-ld: skipping incompatible /nix/store/c79djpxhga419drlbf6v850pvpjfqsgl-gobject-introspection-wrapped-1.72.0/lib/libgirepository-1.0.so when searching for -lgirepository-1.0
```

when gi is in buildInputs

```
⋮ 245│/nix/store/s9m5ww62bsyx0jzylndxicvkmdn7x5bj-aarch64-unknown-linux-gnu-binutils-2.38/bin/aarch64-unknown-linux-gnu-ld: skipping incompatibl
e /nix/store/c79djpxhga419drlbf6v850pvpjfqsgl-gobject-introspection-wrapped-1.72.0/lib/libgirepository-1.0.so when searching for -lgirepository-1.0
```

jtojnar replied:
But [`libgirepostiory`](https://gi.readthedocs.io/en/latest/writingbindings/libgirepository.html), as I understand it, is for runtime loading of typelibs. I still do not see any reason for libraries just providing typelibs to need it – it should only be necessary for stuff like `pygobject`.

Artturin committed:
gobject-introspection: cross improvements

gobject-introspection: copy $dev/share/gobj.../tests when cross-compiling

gobject-introspection: add giscanner-ignore-error-return-codes-from-ldd-wrapper patch

polkit-aarch64-unknown-linux-gnu> [102/104] Generating src/polkitagent/PolkitAgent-1.0.gir with a custom command
polkit-aarch64-unknown-linux-gnu> FAILED: src/polkitagent/PolkitAgent-1.0.gir
polkit-aarch64-unknown-linux-gnu> <storepath>-gobject-introspection-wrapped-1.72.0-dev/bin/g-ir-scanner --no-libtool --namespace=PolkitAgent --nsversion=1.0 --warn-all --output src/polkitagent/PolkitAgent-1.0.gir --c-include=polkitagent/polkitagent.h -
D_POLKIT_AGENT_COMPILATION -D_POLKIT_COMPILATION -I/build/source/src/polkitagent -I/build/source/build/src/polkitagent -I/build/source/. -I/build/source/build/. -I/build/source/src/. -I/build/source/build/src/. --filelist=/build/source/build/src/polkitagent/libpolkit-agent-1.so.0.0.0
.p/PolkitAgent_1.0_gir_filelist --include=Gio-2.0 --include-uninstalled=src/polkit/Polkit-1.0.gir --pkg-export=polkit-agent-1 --cflags-begin -DHAVE_CONFIG_H -I/build/source/. -I/build/source/build/. -I/build/source/src/. -I/build/source/build/src/. -I/nix/store/lqncny8acp6hkclsgbfnrb
65i9hscsmn-expat-aarch64-unknown-linux-gnu-2.4.8-dev/include -I<storepath>-glib-aarch64-unknown-linux-gnu-2.72.2-dev/include/gio-unix-2.0 -I<storepath>-glib-aarch64-unknown-linux-gnu-2.72.2-dev/include -I/nix/store/qm1wa
x4c7sgf5h3b6dhspxb5miayjyc9-glib-aarch64-unknown-linux-gnu-2.72.2-dev/include/glib-2.0 -I<storepath>-glib-aarch64-unknown-linux-gnu-2.72.2/lib/glib-2.0/include -I<storepath>-gobject-introspection-aarch64-unknown-linux-gn
u-1.72.0-dev/include/gobject-introspection-1.0 --cflags-end --add-include-path=<storepath>-gobject-introspection-aarch64-unknown-linux-gnu-1.72.0-dev/share/gir-1.0 --add-include-path=/build/source/build/src/polkit -L/build/source/build/src/polkitagent
--library polkit-agent-1 -L/build/source/build/src/polkit -L<storepath>-expat-aarch64-unknown-linux-gnu-2.4.8/lib --extra-library=expat -L<storepath>-glib-aarch64-unknown-linux-gnu-2.72.2/lib --extra-library=gio-2.0 --ex
tra-library=gobject-2.0 --extra-library=glib-2.0 -L<storepath>-gobject-introspection-aarch64-unknown-linux-gnu-1.72.0/lib --extra-library=girepository-1.0 --sources-top-dirs /build/source/subprojects/ --sources-top-dirs /build/source/build/subprojects/
polkit-aarch64-unknown-linux-gnu> g-ir-scanner: link: aarch64-unknown-linux-gnu-gcc -o /build/source/build/tmp-introspectzp2ldkyk/PolkitAgent-1.0 /build/source/build/tmp-introspectzp2ldkyk/PolkitAgent-1.0.o -L. -Wl,-rpath,. -Wl,--no-as-needed -L/build/source/build/src/polkitagent -Wl
,-rpath,/build/source/build/src/polkitagent -L/build/source/build/src/polkit -Wl,-rpath,/build/source/build/src/polkit -L<storepath>-expat-aarch64-unknown-linux-gnu-2.4.8/lib -Wl,-rpath,<storepath>-expat-aarch64-unknown-
linux-gnu-2.4.8/lib -L<storepath>-glib-aarch64-unknown-linux-gnu-2.72.2/lib -Wl,-rpath,<storepath>-glib-aarch64-unknown-linux-gnu-2.72.2/lib -L<storepath>-gobject-introspection-aarch64-unk
nown-linux-gnu-1.72.0/lib -Wl,-rpath,<storepath>-gobject-introspection-aarch64-unknown-linux-gnu-1.72.0/lib -lpolkit-agent-1 -lexpat -lgio-2.0 -lgobject-2.0 -lglib-2.0 -lgirepository-1.0 -L<storepath>-glib-aarch64-unknow
n-linux-gnu-2.72.2/lib -lgio-2.0 -lgobject-2.0 -Wl,--export-dynamic -lgmodule-2.0 -pthread -lglib-2.0
polkit-aarch64-unknown-linux-gnu> Traceback (most recent call last):
polkit-aarch64-unknown-linux-gnu>   File "<storepath>-gobject-introspection-wrapped-1.72.0-dev/bin/.g-ir-scanner-wrapped", line 99, in <module>
polkit-aarch64-unknown-linux-gnu>     sys.exit(scanner_main(sys.argv))
polkit-aarch64-unknown-linux-gnu>   File "<storepath>-gobject-introspection-wrapped-1.72.0/lib/gobject-introspection/giscanner/scannermain.py", line 646, in scanner_main
polkit-aarch64-unknown-linux-gnu>     shlibs = create_binary(transformer, options, args)
polkit-aarch64-unknown-linux-gnu>   File "<storepath>-gobject-introspection-wrapped-1.72.0/lib/gobject-introspection/giscanner/scannermain.py", line 471, in create_binary
polkit-aarch64-unknown-linux-gnu>     shlibs = resolve_shlibs(options, binary, options.libraries)
polkit-aarch64-unknown-linux-gnu>   File "<storepath>-gobject-introspection-wrapped-1.72.0/lib/gobject-introspection/giscanner/shlibs.py", line 179, in resolve_shlibs
polkit-aarch64-unknown-linux-gnu>     _resolve_non_libtool(options, binary, non_libtool))
polkit-aarch64-unknown-linux-gnu>   File "<storepath>-gobject-introspection-wrapped-1.72.0/lib/go
bject-introspection/giscanner/shlibs.py", line 112, in _resolve_non_libtool
polkit-aarch64-unknown-linux-gnu>     output = subprocess.check_output(args)
polkit-aarch64-unknown-linux-gnu>   File "<storepath>-python3-3.10.5/lib/python3.10/subprocess.py", line 420, in check_output
polkit-aarch64-unknown-linux-gnu>     return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
polkit-aarch64-unknown-linux-gnu>   File "<storepath>-python3-3.10.5/lib/python3.10/subprocess.py", line 524, in run
polkit-aarch64-unknown-linux-gnu>     raise CalledProcessError(retcode, process.args,
polkit-aarch64-unknown-linux-gnu> subprocess.CalledProcessError: Command '['<storepath>-prelink-unstable-2019-06-24/bin/prelink-rtld', '/build/source/build/tmp-introspectzp2ldkyk/PolkitAgent-1.0']' returned non-zero exit status 127.
polkit-aarch64-unknown-linux-gnu> [103/104] Generating src/polkit/Polkit-1.0.typelib with a custom command
polkit-aarch64-unknown-linux-gnu> ninja: build stopped: subcommand failed.

exit status 127

gobject-introspection: this should fix pkgsMusl pkgsStatic

by not putting `/nix/store/iw35xj69w190w7vw17l8mac8srg0vpyz-gobject-introspection-static-x86_64-unknown-linux-musl-1.72.0.drv`

in their dependency tree when gobject-introspection is in
nativeBuildInputs ex .#pkgsStatic.libjxl

gobject-introspection: pc file: use binaries from path instead of a abs path to them

fixes issues in networkmanager and some other packages
```
gi._error.GError: g-invoke-error-quark: Could not locate g_option_error_quark: /nix/store/...-glib-aarch64-unknown-linux-gnu-2.72.2/lib/libglib-2.0.so.0:
cannot open shared object file: No such file or directory (1)
```

see comment

gobject-introspection: propagate gobject-introspection-unwrapped in wrapper

with this it is no longer needed to put gobject-introspection in
buildInputs in addition to nativeBuildInputs

Artturin commented:
draft because im still working on this

Artturin committed:
gobject-introspection: use targetOffset to look for libraries for target

fixes it finding the build and target gobject-introspection when it should only find the
target gobject-introspection

Artturin committed:
librdf_raptor2: 2.0.15 -> unstable-2022-06-06

switches to using cmake and fixes cross-compilation

Artturin commented on pkgs/development/libraries/tracker/default.nix:
```diff
@@ -50,11 +56,8 @@ stdenv.mkDerivation rec {
-    dbus # used for checks and pkg-config to install dbus service/s
-  ] ++ checkInputs; # gi is in the main meson.build and checked regardless of
-                    # whether tests are enabled
+    (buildPackages.python3.withPackages (p: [ p.pygobject3 ]))
```
can be switched to use pythonForBuild once https://github.com/NixOS/nixpkgs/pull/177888 is merged

SuperSandro2000 reviewed (commented):
Can we set strictDeps to prevent regressions?

Artturin commented on pkgs/development/libraries/libical/default.nix:
```diff
@@ -37,22 +36,21 @@ stdenv.mkDerivation rec {
   installCheckInputs = [
     # running libical-glib tests
-    (python3.withPackages (pkgs: with pkgs; [
+    (buildPackages.python3.withPackages (pkgs: with pkgs; [
```
same as tracker

Artturin commented:
> Can we set strictDeps to prevent regressions?

First fix then https://github.com/NixOS/nixpkgs/pull/177538#issuecomment-1154264286

Artturin commented:
Many meson projects are looking for gir, vapi, and gtk-doc wrong
See https://gitlab.gnome.org/GNOME/libgudev/-/merge_requests/27 for how to fix it

jtojnar commented on pkgs/tools/networking/networkmanager/default.nix:
```diff
@@ -61,7 +64,8 @@ stdenv.mkDerivation rec {
 
-  outputs = [ "out" "dev" "devdoc" "man" "doc" ];
+  outputs = [ "out" "dev" ]
+    ++ lib.optionals (stdenv.buildPlatform == stdenv.hostPlatform) [ "devdoc" "man" "doc" ];
```
Are the man pages not cross buildable either?

Artturin replied:
This patch will be needed https://github.com/void-linux/void-packages/blob/master/srcpkgs/NetworkManager/patches/no-gtk-doc.patch

Though I can just copy the outputs from the native package to the new one

jtojnar commented on pkgs/tools/networking/networkmanager/default.nix:
```diff
@@ -140,10 +146,15 @@ stdenv.mkDerivation rec {
     newt
     libsoup
     jansson
+    dbus
```
Add a comment that this is only used for getting service installation directory paths from the pkg-config, since libdbus should not really be needed.

Artturin committed:
mesonEmulatorHook: check if the target binaries can be executed

this prevents having to bring in the emulator when compiling e.g. pkgsStatic

Artturin committed:
gst_all_1: don't disable gobject-introspection when cross

Artturin committed:
python310Packages.pygobject3: fix cross

Artturin committed:
libgudev: dont specialcase cross and pull patch

to not have to have vala in both nativeBuildInputs and buildInputs

https://gitlab.gnome.org/GNOME/libgudev/-/merge_requests/27

> build: let meson handle gir, vala, gtk-doc dependencies
> gnome.generate_gir
> gnome.generate_vapi
> gnome.gtkdoc
>
> all do their own dependency finding in a cross-compilation friendly manner while
> using dependency() with the default options does not
>
> fixes finding vapigen while cross-compiling

Artturin committed:
networkmanager: fix cross

copy devdoc and man from native to cross
doc does not have to be copied

Artturin committed:
various: enable gobject-introspection when cross-compiling

and fix cross in some

Mindavi commented on pkgs/development/libraries/gdk-pixbuf/default.nix:
```diff
@@ -108,9 +105,9 @@ stdenv.mkDerivation rec {
+    '' + ''
       # We need to install 'loaders.cache' in lib/gdk-pixbuf-2.0/2.10.0/
-      $dev/bin/gdk-pixbuf-query-loaders --update-cache
+      ${stdenv.hostPlatform.emulator buildPackages} $dev/bin/gdk-pixbuf-query-loaders --update-cache
```
Meh, too bad we have to do this :(

Mindavi commented on pkgs/development/libraries/libgudev/default.nix:
```diff
@@ -22,6 +22,15 @@ stdenv.mkDerivation rec {
   };
 
+  patches = [
+    # https://gitlab.gnome.org/GNOME/libgudev/-/merge_requests/27
```
Could use some love 😉

Artturin replied:
done and updated patch

Artturin replied:
done

Mindavi commented on pkgs/development/libraries/librdf/raptor2.nix:
```diff
@@ -1,26 +1,28 @@
 stdenv.mkDerivation rec {
   pname = "raptor2";
-  version = "2.0.15";
+  version = "unstable-2022-06-06";
```
Great that you managed to fix this too!

Mindavi commented on pkgs/development/libraries/opencv/4.x.nix:
```diff
@@ -291,11 +292,15 @@ stdenv.mkDerivation {
+    # "OpenCV disables pkg-config to avoid using of host libraries. Consider using PKG_CONFIG_LIBDIR to specify target SYSROOT"
+    # but we have proper separation of build and host libs :), fixes cross
+    "-DOPENCV_ENABLE_PKG_CONFIG=ON"
     (opencvFlag "IPP" enableIpp)
```
Woo!

Mindavi reviewed (approved):
One question and mostly some random comments and thoughts. Thanks for moving this forward! Also nice that we can get rid of gi in buildInputs in a few packages.

Mindavi commented:
Btw, haven't tested building this, might do that but I don't have that much time right now. We'll see.

Mindavi commented on pkgs/development/libraries/gobject-introspection/wrapper.nix:
```diff
@@ -7,9 +7,13 @@
+gobject-introspection-unwrapped.overrideAttrs (previousAttrs: {
   pname = "gobject-introspection-wrapped";
-  postFixup = ''
+  depsTargetTargetPropagated = [ gobject-introspection-unwrapped ];
```
Ahhh, then it's quite important to get that working too for the full experience. But fine if we defer that, doesn't make sense to do it in this pr

jtojnar replied:
Last time I checked, the meson build was still incomplete but feel free to try it.

Artturin replied:
> Last time I checked, the meson build was still incomplete but feel free to try it.

at least arch uses the meson builds for both

Artturin replied:
also, we will need this for `build == host == target` when `strictDeps = true` if we do not add gobject-introspection to buildInputs

``
  depsBuildBuild = [
    pkg-config
  ];
```

Artturin replied:
btw this is the error with wich the freedesktop modem related packages fail with when not using meson

```
libqmi-aarch64-unknown-linux-gnu> /build/libqmi-1.30.8/src/libqmi-glib/tmp-introspect2d8e4td2/.libs/Qmi-1.0: error while loading shared libraries: libqmi-glib.so.5: cannot open shared object file: No such file or directory
libqmi-aarch64-unknown-linux-gnu> Command '['/nix/store/wr4mjr9gadvwz6c88fgz4ldzn21sxkpz-qemu-7.0.0/bin/qemu-aarch64', '/build/libqmi-1.30.8/src/libqmi-glib/tmp-introspect2d8e4td2/.libs/Qmi-1.0', '--introspect-dump=/build/libqmi-1.30.8/src/libqmi-glib/tmp-introspect2d8e4td2/functions.txt,/build/libqmi-1.30.8/src/libqmi-glib/tmp-introspect2d8e4td2/dump.xml']' returned non-zero exit status 127.
libqmi-aarch64-unknown-linux-gnu> make[4]: *** [/nix/store/jwpcacx8sd4vj0ckk5wxrzn2iwcpp35s-gobject-introspection-aarch64-unknown-linux-gnu-1.72.0-dev/share/gobject-introspection-1.0/Makefile.introspection:156: Qmi-1.0.gir] Error 1
```

Artturin commented on pkgs/development/libraries/gobject-introspection/default.nix:
```diff
@@ -27,7 +27,7 @@
 # In that case its about 6MB which could be separated
 
-stdenv.mkDerivation rec {
+stdenv.mkDerivation (finalAttrs: {
```
see postFixup

Artturin commented on pkgs/development/libraries/gobject-introspection/default.nix:
```diff
@@ -120,11 +124,21 @@ stdenv.mkDerivation rec {
+  # when cross-compiling and using the wrapper then when a package looks up the g_ir_X
+  # variable with pkg-config they'll get the host version which can't be run
+  # switch the variables to use g_ir_X from path instead of an absolute path
+  postFixup = lib.optionalString (!lib.hasSuffix "wrapped" finalAttrs.pname) ''
```
i think it is fine the way it is

Mindavi replied:
let's leave it then

Artturin commented on pkgs/development/libraries/gobject-introspection/setup-hook.sh:
```diff
@@ -10,7 +10,7 @@ make_gobject_introspection_find_gir_files() {
 }
 
-addEnvHooks "$hostOffset" make_gobject_introspection_find_gir_files
+addEnvHooks "$targetOffset" make_gobject_introspection_find_gir_files
```
@Ericson2314 also had it in his PR https://github.com/NixOS/nixpkgs/pull/88222/files#diff-c0d47b1d6e8a614ceb2b7083f7ebc37d288cbfc1fac81ec83ccde8a94718eb8c

SuperSandro2000 commented:
> list of files in which strictDeps will be enabled in separate PRs

Yes, please. This PR is getting quite big.

SuperSandro2000 commented on pkgs/development/python-modules/pygobject/3.nix:
```diff
@@ -37,8 +42,9 @@ buildPythonPackage rec {
 
   buildInputs = [
-    glib
+    # # .so files link to these
```
```suggestion
    # .so files link to these
```

Mindavi commented:
Hey Artturin, can we get this in soon rather than piling up stuff? I don't think it's very convenient to keep adding things, we can always iterate on this later on, right?

Especially the more leafy fixes are fine to add later, IMO.

Also cc @Ericson2314 who may be interested in this.

MatthewCroughan reviewed (approved):
This fixes quite a lot of issues for me personally when trying to cross-compile for riscv64, so I'd really love to see this merged.

Artturin commented:
Modemmanager (a dep of networkmanager) should be fixed before merge

I'll do it tomorrow as I was on vacation for 3 days

Artturin commented:
Also strictDeps gobject-introspection native will be broken due to not propagating itself to buildInputs while in nativeBuildInputs when not specified in buildInputs

Artturin committed:
gobject-introspection: do not propagate target gobject-introspection in wrapper

broke modemmanager and some of its deps

libqmi-aarch64-unknown-linux-gnu>   GISCAN   Qmi-1.0.gir
libqmi-aarch64-unknown-linux-gnu> /build/libqmi-1.30.8/src/libqmi-glib/tmp-introspect2d8e4td2/.libs/Qmi-1.0: error while loading shared libraries: libqmi-glib.so.5: cannot open shared object file: No such file or directory
libqmi-aarch64-unknown-linux-gnu> Command '['/nix/store/wr4mjr9gadvwz6c88fgz4ldzn21sxkpz-qemu-7.0.0/bin/qemu-aarch64', '/build/libqmi-1.30.8/src/libqmi-glib/tmp-introspect2d8e4td2/.libs/Qmi-1.0', '--introspect-dump=/build/libqmi-1.30.8/src/libqmi-glib/tmp-introspect2d8e4td2/functions.txt,/build/libqmi-1.30.8/src/libqmi-glib/tmp-introspect2d8e4td2/dump.xml']' returned non-zero exit status 127.
libqmi-aarch64-unknown-linux-gnu> make[4]: *** [/nix/store/jwpcacx8sd4vj0ckk5wxrzn2iwcpp35s-gobject-introspection-aarch64-unknown-linux-gnu-1.72.0-dev/share/gobject-introspection-1.0/Makefile.introspection:156: Qmi-1.0.gir] Error 1

Artturin commented:
reverted the propagation 

original commits can be found in https://github.com/Artturin/nixpkgs/tree/fixcross3-backup

Artturin committed:
various: readd gobject-introspection to buildInputs

see eb829035c938022bb064faef56a7b1346443e034

Artturin commented:
this flake built
```
{
  description = "cross-compile the sd-image-aarch64 on x86_64-linux to aarch64-multiplatform";
  inputs = {
    #nixpkgs.url = "github:NixOS/nixpkgs/nixos-unstable";
    nixpkgs.url = "/home/artturin/nixgits/my-nixpkgs/.worktree/1";
  };

  outputs = { self, nixpkgs }: {

    nixosConfigurations.cross-vm = nixpkgs.lib.nixosSystem {
      system = "x86_64-linux";
      modules = [
        ({ pkgs, lib, modulesPath, ... }: {
          nixpkgs = {
            crossSystem = lib.systems.examples.aarch64-multiplatform;
          };

          imports = [ (modulesPath + "/installer/sd-card/sd-image-aarch64.nix") ];

          users.mutableUsers = false;
          users.users.root = {
            password = "root";
          };
          users.users.user = {
            password = "user";
            isNormalUser = true;
            extraGroups = [ "wheel" ];
          };
          system.stateVersion = "22.05";
        })
      ];
    };

    packages.x86_64-linux.default = self.nixosConfigurations.cross-vm.config.system.build.sdImage;
    packages.x86_64-linux.vm = self.nixosConfigurations.cross-vm.config.system.build.vm;
    apps.x86_64-linux.vm = {
      type = "app";
      program = "${self.defaultPackage.x86_64-linux}/bin/run-nixos-vm-cross";
    };
  };
}
```

trofi commented:
Bisect claims that `gtk2` broke after a83d2f7c918a5badab204d57d8620adb926e618b `gobject-introspection: cross improvements`:

```
$ nix build -f. gtk2 -L
...
(lost logs will update in a ~hour)
```

Artturin commented:
`gtk+> make[4]: *** No rule to make target 'g-ir-scanner', needed by 'Gtk-2.0.gir'.  Stop.`

Artturin commented:
fixed in https://github.com/NixOS/nixpkgs/pull/181413

trofi commented:
Bisect claims 10362d728084d8810f6af2f8588b172e092b01e6 `librdf_raptor2: 2.0.15 -> unstable-2022-06-06` boroke a bunch of packages in `staging-next`. All have the same symptom: underlinked `liblrdf.so` library is not usable by it's final depends:

Broken packages are `non`, `tony`, `rascal`, `hydrogen`, `guitarix`. Example failure:

```
$ nix build -f. non -L
...
non-unstable> /nix/store/ydvvnn6mi99kyp9an1gnrcga1dr5v5x0-binutils-2.38/bin/ld: /nix/store/1kkgkg3irmk78pk6c96154fns1k5133g-lrdf-0.6.1/lib/liblrdf.so: undefined reference to `xsltNewSecurityPrefs'
non-unstable> /nix/store/ydvvnn6mi99kyp9an1gnrcga1dr5v5x0-binutils-2.38/bin/ld: /nix/store/1kkgkg3irmk78pk6c96154fns1k5133g-lrdf-0.6.1/lib/liblrdf.so: undefined reference to `xmlXPathNewContext'
```

I think `lrdf` library is linked incorrectly and is missing some `xml` library.

The simplest way to trigger the bug earlier is to add an `LDFLAGS = "-Wl,--no-undefined"` hack locally:

```
$ nix build --impure --expr 'with import ./.{}; lrdf.overrideAttrs (oa: { LDFLAGS = "-Wl,--no-undefined"; })' -L
...
lrdf> libtool: link: gcc -shared  -fPIC -DPIC  .libs/lrdf.o .libs/lrdf_multi.o .libs/md5.o   -L/nix/store/28xxhxqb97j48yj4ivvl0c6kih9l2r52-raptor2-unstable-2022-06-06/lib -lraptor2  -g -g -O2 -g -O2 -Wl,--no-undefined   -Wl,-soname -Wl,liblrdf.so.2 -o .libs/liblrdf.so.2.0.0
lrdf> /nix/store/ydvvnn6mi99kyp9an1gnrcga1dr5v5x0-binutils-2.38/bin/ld: /nix/store/28xxhxqb97j48yj4ivvl0c6kih9l2r52-raptor2-unstable-2022-06-06/lib/libraptor2.a(raptor_sax2.c.o): in function `raptor_sax2_parse_chunk':
lrdf> (.text+0x511): undefined reference to `xmlParseChunk'
lrdf> /nix/store/ydvvnn6mi99kyp9an1gnrcga1dr5v5x0-binutils-2.38/bin/ld: (.text+0x55d): undefined reference to `xmlCreatePushParserCtxt'
lrdf> /nix/store/ydvvnn6mi99kyp9an1gnrcga1dr5v5x0-binutils-2.38/bin/ld: (.text+0x589): undefined reference to `xmlCtxtUseOptions'
```

trofi commented:
> Bisect claims [10362d7](https://github.com/NixOS/nixpkgs/commit/10362d728084d8810f6af2f8588b172e092b01e6) `librdf_raptor2: 2.0.15 -> unstable-2022-06-06` boroke a bunch of packages in `staging-next`. All have the same symptom: underlinked `liblrdf.so` library is not usable by it's final depends:

> I think `lrdf` library is linked incorrectly and is missing some `xml` library.

Possible fix (or workaround) is to enable shared libs by default. Proposed as: https://github.com/NixOS/nixpkgs/pull/182727
